lightning/src/routing/router.rs

```diff
@@ -328,7 +328,7 @@ fn compute_fees(amount_msat: u64, channel_fees: RoutingFees) -> Option<u64> {
 /// The fees on channels from us to next-hops are ignored (as they are assumed to all be
 /// equal), however the enabled/disabled bit on such channels as well as the
 /// htlc_minimum_msat/htlc_maximum_msat *are* checked as they may change based on the receiving node.
-pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, payee: &PublicKey, first_hops: Option<&[&ChannelDetails]>,
+pub fn get_route<L: Deref>(our_node_id: &PublicKey, network: &NetworkGraph, payee: &PublicKey, payee_features: Option<InvoiceFeatures>, first_hops: Option<&[&ChannelDetails]>,
```

Comment:
High-level question on how this should be used:

Rather than passing in `None` in a ton of places, would it be preferable for the caller to modify the returned route? Or maybe instead provide some sort of wrapper that given an `InvoiceFeatures ` calls `get_route` and updates the returned value as necessary?

Reply:
Hmm, I strongly think its not preferable to have the user modify the `Route` object. While its all visible, it really shouldn't fall to the user to figure out how to tweak it properly. If you're worried about the number of arguments, we could pull them out into an args struct, but the types/argument names are all pretty clear.

Reply:
I was leaning more to the second option of adding something like `get_route_using_features` which delegates to `get_route` and then updates the result before returning. Then `get_route` doesn't need an extra parameter and all those call sites taking `None`, which isn't very readable (_what is `None` for?_).

Reply:
In this case the None-or-not also impacts how we build the route (MPP or not), so we can't just drop it and copy it into the route at the end.

Reply:
Not sure I follow. Isn't lines 969-974 doing just that?

Reply:
Yes, but lines 402-411 is also checking the features object provided to decide if we can use MPP.

Reply:
Ah, that's what I get for looking at just one commit. :P I still wonder if we could come up with a better API, which would probably easier to do if `get_route` was refactor into smaller functions. I feel like passing `None` in places is similar to passing a `bool`. Just makes reading code difficult.

Reply:
No need to hold up on this, BTW. We could take a more wholistic look at the APIs later.

Reply:
Yea, generally agree, it definitely makes some tests harder. That said, in "real" applications it'd be somewhat "wrong" to pass None in any of the fields - the last-hop hints and invoice features should be at least be passed if they're present.